### ISSUE TITLE
Mobilenet V2 classification model in cv2 dnn

### ISSUE THREAD

**Issue by VenkatKrish17** [opencv/opencv#16261]
##### System information (version)

Opencv => 4.0.0 (C++ custom build, includes dnn module)
Operating system => Ubuntu
Compiler => GCC 
##### Summary 
I have a transfer learning mobilenetV2 classification model, which I want to use in a C++ code using the opencv's cv2 dnn module. 

**The simplest way -** 
I have a custom trained mobilenetV2 classification model's frozen inference graph. 
From the frozen inference graph, I generated the .pbtxt file using cv2.dnn.writeTextGraph method and ran the prediction using cv2::dnn::readNetFromTensorflow method 

With just the model file reference alone got the below error. 

> OpenCV(4.0.0) /home/venkat/OPENCV4/opencv/modules/dnn/src/tensorflow/tf_importer.cpp:497: error: (-2:Unspecified error) Input layer not found: module_apply_default/MobilenetV2/expanded_conv/output in function 'connect'

with both the model file and the config file got the below error.

>   what():  OpenCV(4.0.0) /home/venkat/OPENCV4/opencv/modules/dnn/src/tensorflow/tf_importer.cpp:616: error: (-215:Assertion failed) const_layers.insert(std::make_pair(name, li)).second in function 'addConstNodes'

**The complex way**

I followed the instructions mentioned in 
https://github.com/opencv/opencv/issues/11577#issuecomment-391372527  
and 
https://github.com/opencv/opencv/issues/11452
and 
https://github.com/opencv/opencv/issues/12075

All the above instructions are typically for the object detection models. 
I looked at the https://github.com/opencv/opencv/blob/master/samples/dnn/tf_text_graph_ssd.py which is for mobilenet ssd model. 

So from various above sources, I ended up creating the below code(in python) for converting the model to be compatible with cv2 dnn

with tf.gfile.FastGFile('mobilenet_v2_100_224.pb', 'rb') as f:
    graph_def = tf.GraphDef()
    graph_def.ParseFromString(f.read())

    graph_def = optimize_for_inference_lib.optimize_for_inference(graph_def, ['final_retrain_ops/biases/final_biases'], ['final_result'], tf.float32.as_datatype_enum)
    graph_def = TransformGraph(graph_def, ['final_retrain_ops/biases/final_biases'], ['final_result'], ['remove_nodes(op=PlaceholderWithDefault)','sort_by_execution_order'])

    for i in reversed(range(len(graph_def.node))):
        #code blocks 1
        if graph_def.node[i].op == 'Const':
            del graph_def.node[i]
        #code block 1 ends
        for attr in ['T', 'data_format', 'Tshape', 'N', 'Tidx', 'Tdim',
                    'use_cudnn_on_gpu', 'Index', 'Tperm', 'is_training',
                    'Tpaddings']:
            if attr in graph_def.node[i].attr:
                print("Deleting")
                del graph_def.node[i].attr[attr]
        print(len(graph_def.node))
    unconnected_nodes = []
    for node in graph_def.node:
        unconnected_nodes.append(node.name)
        for inp in node.input:
            if inp in unconnected_nodes:
                unconnected_nodes.remove(inp)
    unconnected_nodes.remove("final_result")
    for name in unconnected_nodes:
        for i in range(len(graph_def.node)):
            if graph_def.node[i].name == name:
                # print(name)
                del graph_def.node[i]
                break
    print("leftover nodes ")
    for node in graph_def.node:
        print(node.name)

    with tf.gfile.FastGFile('graph.pb', 'wb') as f:
            f.write(graph_def.SerializeToString())
    cv2.dnn.writeTextGraph('graph.pb','graph.pbtxt')

With the graph.pb and graph.pbtxt when I tried to load the model, I encountered the below error. 

When I delete the "Const" op nodes (code block 1 in the above code, commented accordingly), I get the following error while writing the text graph. 

> error: (-215:Assertion failed) nodesMapIt != nodesMap.end() in function 'sortByExecutionOrder'

So I comment the code block 1, so my graph gets generated,
with the below log

> ...
> W1230 17:47:58.718445 139948826896192 optimize_for_inference_lib.py:244] Didn't find expected Conv2D input to 'module_apply_default/MobilenetV2/expanded_conv_16/depthwise/BatchNorm/FusedBatchNorm'
> 2019-12-30 17:48:00.034702: I tensorflow/tools/graph_transforms/transform_graph.cc:317] Applying remove_nodes
> 2019-12-30 17:48:00.041844: I tensorflow/tools/graph_transforms/transform_graph.cc:317] Applying sort_by_execution_order

 and with the generated graph, when I tried to load in cv2 dnn module, ended up at this error. 

> error: (-215:Assertion failed) const_layers.insert(std::make_pair(name, li)).second in function 'addConstNodes'

I also tried using the https://github.com/opencv/opencv/blob/master/samples/dnn/tf_text_graph_common.py file to convert, still, I ended up in errors. 

Would like to know more on this issue, and the proper way to convert a custom trained mobile net v2 classification model compatible for opencv dnn module.

**Comment by VenkatKrish17** [comment]
@dkurt  I have seen your replies in most of the above issues thread I mentioned. I wish you can help me also to fix the issue. 
Thanks.

**Comment by dkurt** [comment]
> Opencv => 4.0.0 (C++ custom build, includes dnn module)

Have you tried the latest version?

> with tf.gfile.FastGFile('mobilenet_v2_100_224.pb', 'rb') as f:

This is not object detection SSD model that means the script tf_text_graph_ssd.py is not applicable for it. Can you share a model itself?

**Comment by VenkatKrish17** [comment]
Do you mean the latest version of tensorflow or opencv ?  
I use tensorflow 1.14.

**Comment by dkurt** [comment]
@VenkatKrish17, The latest version of OpenCV, yes

**Comment by VenkatKrish17** [comment]
I have 4.1.2 for my python environment. 
When i tried converting the graph using cv2.writeTextGraph, I was able to generate the pbtxt file. 
But when I used readNetFromTensorflow, it gave me the same below error. 

OpenCV(4.1.2) /io/opencv/modules/dnn/src/tensorflow/tf_importer.cpp:663: error: (-215:Assertion failed) const_layers.insert(std::make_pair(name, li)).second in function 'addConstNodes'

Do you really want me to try with 4.2.3 which is released last week? Does that have changes related to dnn module ?

**Comment by dkurt** [comment]
@VenkatKrish17, please upload mentioned network somewhere so we can reproduce the problem.

**Comment by VenkatKrish17** [comment]
@dkurt  i have emailed you the .pb file to dmitry.kurtaev+github@gmail.com since its private cannot post the link here.

**Comment by VenkatKrish17** [comment]
closing this issue, as the model itself is an issue. 
I simply trained the model in keras, and used this script (https://www.dlology.com/blog/how-to-convert-trained-keras-model-to-tensorflow-and-make-prediction/)  to convert it to pb. (Without optimising for inference)
Then loaded the pb, it worked fine.
